sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala

```diff
@@ -260,11 +260,11 @@ private[sql] class DataFrameImpl protected[sql](
 
   override def take(n: Int): Array[Row] = head(n)
 
-  override def collect(): Array[Row] = queryExecution.executedPlan.executeCollect()
+  override def collect(): Array[Row] = rdd.collect()
 
   override def collectAsList(): java.util.List[Row] = java.util.Arrays.asList(rdd.collect() :_*)
 
-  override def count(): Long = groupBy().count().rdd.collect().head.getLong(0)
+  override def count(): Long = rdd.count()
```

Comment:
@marmbrus is correct. rdd.count() doesn't go through the optimizer. The original solution goes through the optimizer.

Maybe a better change is to add some inline comment to explain this makes sure it goes through the optimizer, etc.